clang/test/SemaSYCL/max-concurrency-ast.cpp

```diff
@@ -0,0 +1,94 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -Wno-sycl-2017-compat -ast-dump %s | FileCheck %s
+
+// Tests for AST of Intel FPGA max concurrency function attribute.
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+queue q;
+
+// CHECK: FunctionDecl {{.*}} func1 'void ()'
+// CHECK-NEXT: CompoundStmt
+// CHECK-NEXT: SYCLIntelFPGAMaxConcurrencyAttr
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 1
+// CHECK-NEXT: IntegerLiteral {{.*}} 'int' 1
+[[intel::max_concurrency(1)]] void func1() {}
+
+// CHECK: FunctionDecl {{.*}} func2 'void ()'
+// CHECK-NEXT: CompoundStmt
+// CHECK-NEXT: SYCLIntelFPGAMaxConcurrencyAttr
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 0
+// CHECK-NEXT: IntegerLiteral {{.*}} 'int' 0
+[[intel::max_concurrency(0)]] void func2() {}
+
+// CHECK: FunctionTemplateDecl {{.*}} func3
+// CHECK: FunctionDecl {{.*}} func3 'void ()'
+// CHECK-NEXT: CompoundStmt
+// CHECK_NEXT: SYCLIntelFPGAMaxConcurrencyAttr
+// CHECK_NEXT: DeclRefExpr {{.*}} 'int' NonTypeTemplateParm {{.*}} 'N' 'int'
+// CHECK: FunctionDecl {{.*}} func3 'void ()'
+// CHECK-NEXT: TemplateArgument integral 5
+// CHECK-NEXT: CompoundStmt
+// CHECK-NEXT: SYCLIntelFPGAMaxConcurrencyAttr
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 5
+// CHECK-NEXT: SubstNonTypeTemplateParmExpr
+// CHECK-NEXT: NonTypeTemplateParmDecl
+// CHECK-NEXT: IntegerLiteral {{.*}} 'int' 5
+template <int N>
+[[intel::max_concurrency(N)]] void func3() {}
+
+class KernelFunctor {
+public:
+  void operator()() const {
+    func1();
+    func2();
+  }
+};
+
+template <int N>
+class KernelFunctor2 {
+public:
+  [[intel::max_concurrency(N)]] void operator()() const {
+  }
+};
+
+void foo() {
+  q.submit([&](handler &h) {
+    // CHECK: FunctionDecl {{.*}}kernel_name_1
```

Comment:
Can you explain what you are testing here?  Is this attribute supposed to be propagated to kernel or not ?